### DISCUSSION TITLE
Architectural Stability and Production use

### DISCUSSION THREAD

**Discussion by atul161** [JanusGraph/janusgraph#3049] [General]
Hey folks,

I am going to deploy janus graph in production and want to ask , is janus server is stateless if i am using cassandra as a storage and 
elastic for search.

Where did janus graph store schema?

Can i just create stateless pod of janus server and connect it with external storage like cassandra?

Is it good approach?

**Comment by porunov** [comment]
JanusGraph server stores schema inside a storage backend (i.e. Cassandra in your case). 
JanuaGraph servers don't communicate to each other directly but they communicate via the messaging events using a storage backend (Cassandra). JanusGraph uses it to communicate schema updates between all JanusGraph servers. JanusGraph may also acquire id blocks from a storage backend (the default approach) unless you explicitly configure custom ids. Other than that JanusGraph doesn't have any states during queries execution or data loading. JanusGraph nodes are equal (no master nodes, no slave nodes, no coordinator nodes). In case one or more JanusGraph nodes dies it won't affect other JanusGraph nodes. You can scale up or down JanusGraph very easily. Just add more nodes when you need more parallel computation power or remove JanusGraph nodes when you need less.
Starting a JanusGraph server and connecting it to Cassandra is a good approach. Another approach could be using JanusGraph on application level if you use Java (it will skip serialization and additional network hop). Both approaches are good for their own use cases.

**Reply by atul161** [reply]
Thanks for clarifying!!

We have a use case where we will deploy janus with cassandra to handle billions of linking and having billions of vertex.

I have one more question, can we write schema file explicitly without going into console?

I mean whenever janus graph will start i want first it create data model if not created or not exist.

**Reply by porunov** [reply]
> We have a use case where we will deploy janus with cassandra to handle billions of linking and having billions of vertex.

JanusGraph should handle it with Cassandra without any problem. 

> I have one more question, can we write schema file explicitly without going into console? I mean whenever janus graph will start i want first it create data model if not created or not exist.

That's a fair feature request. 
It's not in JanusGraph yet. I did develop a possibility to define schema in JSON for JanusGraph but the implementation has some limitations. Moreover it's not part of JanusGraph codebase but part of another project codebase. If there is enough demand I may work on it later to move that logic under JanusGraph codebase.
Moreover, @farodin91 was developing GRPC schema management which is part of JanusGraph.

**Reply by atul161** [reply]
> 



> > We have a use case where we will deploy janus with cassandra to handle billions of linking and having billions of vertex.
> 
> JanusGraph should handle it with Cassandra without any problem.
> 
> > I have one more question, can we write schema file explicitly without going into console? I mean whenever janus graph will start i want first it create data model if not created or not exist.
> 
> That's a fair feature request. It's not in JanusGraph yet. I did develop a possibility to define schema in JSON for JanusGraph but the implementation has some limitations. Moreover it's not part of JanusGraph codebase but part of another project codebase. If there is enough demand I may work on it later to move that logic under JanusGraph codebase. Moreover, @farodin91 was developing GRPC schema management which is part of JanusGraph.

Thanks for clarifying,
Yes it will be helpful for devs.
We are deploying janus as a stateless (pod in kubernates) and this feature will be helpful for us .
We dont need to create schema manually whenever pod dies.

Is there any api you guys are providing through which we can send script ?

**Reply by farodin91** [reply]
> Is there any api you guys are providing through which we can send script ? 

The protobuf api currently just support the read part. I hope i'm able to start with the write asap.
